src/engine/Engine.h

```diff
@@ -367,8 +369,301 @@ class Engine {
     }
   }
 
-private:
+  template<typename R, int K>
+  struct newOptionalResult {
+    R operator()(unsigned int resultSize) {
+      (void) resultSize;
+      return R();
+    }
+  };
+
+  template<int K>
+  struct newOptionalResult<std::vector<Id>, K> {
+    std::vector<Id> operator()(unsigned int resultSize) {
+      return vector<Id>(resultSize);
+    }
+  };
+
+
+ template<typename A, typename B, typename R, bool aEmpty, bool bEmpty>
+  static void createOptionalResult(const typename A::value_type* a,
```

Comment:
do `a` and `b` have to be pointers, i.e. can they be null? Otherwise a reference may be nicer

Reply:
They can be null.